### PR TITLE
feat: Enable Bot as Callable Skill

### LINKED ISSUE #696: Call a remote skill
## Task Track

Open Item
- 8/30 Depends on SDK

https://microsoft-my.sharepoint.com/:w:/r/personal/luhan_microsoft_com/_layouts/15/Doc.aspx?sourcedoc=%7B03AFB829-07F7-4764-823A-842531D987DC%7D&file=RemoteSkillFuncSpec.docx&action=default&mobileredirect=true

see microsoft/botbuilder-dotnet#2324

### PR DESCRIPTION
**Pull request by zidaneymar**
## Description
Few steps to enable a bot as callable skill:

1. Register app id and app password both for root bot and sub bot

2. Fill the appsettings.json in root bot and sub bot
Root bot:
 ```
  "MicrosoftAppId": "Root Bot AppID",
  "MicrosoftAppPassword": "Root Bot AppPwd",
  "BotFrameworkSkills": [
    {
      "Id": "EchoSkillBot",
      "AppId": "<Sub Bot's AppId>",
      "SkillEndpoint": "http://localhost:1205/api/messages"
    }
  ]
```
Sub Bot:
 ```
  "MicrosoftAppId": "Sub Bot AppID",
  "MicrosoftAppPassword": "Sub Bot AppPwd",
  "AllowedCallers": []
```

3. Add a new json dialog like this:

```
{
    "$schema": "../../app.schema",
    "$kind": "Microsoft.AdaptiveDialog",
    "triggers": [
        {
            "$kind": "Microsoft.OnUnknownIntent",
            "actions": [
              {
                "$kind": "Microsoft.BeginSkill",
                "targetSkill": {
                  "Id": "EchoSkillBot",
                  "AppId": "SubBotAppId",
                  "SkillEndpoint": "http://localhost:1205/api/messages"
                },
                "SkillHostEndpoint": "http://localhost:3979/", // the endpoint where main bot hosted on
              }
            ]
        }
    ]
}
```

image

## Task Item

fixes #696 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code refactor (non-breaking change which improve code quality, clean up, add tests, etc)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Doc update (document update)

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have functionally tested my change

## Screenshots

Please include screenshots or gifs if your PR include UX changes.

### CODE REVIEW

**Comment by cwhitten** [comment]
Is the password of the parent and child bots always required? I don't believe this constraint exists in the current rev of the BF Skill manifest. @gabog can you comment here?

**Comment by gabog** [comment]
Hi, yes, when talking to skills the parent and child bots need to be configured with valid Azure ID apps and passwords, this is needed to authenticate calls back and forth.
The manifest is a public json file that defines what the bot can do, it does need to specify the bot ID (MS App ID) but not the password.

**Comment by luhan2017** [comment]
@gabog , Could you help review this PR and help confirm whether this implementation is correct?

I have several questions needs to confirm with you.
1. A bot have two controllers, message controller(api/messages) and skill controller(api/skill). In this PR, it seems like when the bot is a callee, it will always reciece request from message controller, and the caller will handle the skill callback by skill controller, is this correct? I remember previously skills are using api/skills to recieve messages.
2. we implemented begin a skill as a custom action instead of built-in action in SDK, does SDK side have a plan to implement a built-in action for beginskill? 
3. As for local testing, is it possible for us to disable the validation part and just don't use the settings like MicrosoftAppID, MicrosoftPassword and AllowedCallers....... it will bring complexity for users to test skills if we need to set MicrosoftAppID, MicrosoftPassword.

File: BotProject/Templates/CSharp/BeginSkillLoader.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+using Microsoft.Bot.Builder.Dialogs.Declarative.Loaders;
+using Microsoft.Bot.Builder.Integration.AspNet.Core.Skills;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Bot.Builder.ComposerBot.Json
+{
+    public class BeginSkillLoader : ICustomDeserializer
+    {
+        private SkillHttpClient skillHttpClient;
```
Comment (gabog) [inline BotProject/Templates/CSharp/BeginSkillLoader.cs]: readonly

File: BotProject/Templates/CSharp/BeginSkillLoader.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+        private ConversationState conversationState;
```
Comment (gabog) [inline BotProject/Templates/CSharp/BeginSkillLoader.cs]: readonly

File: BotProject/Templates/CSharp/SkillConversationIdFactory.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder.Skills;
+using Microsoft.Bot.Schema;
+using Newtonsoft.Json;
+
+namespace Microsoft.Bot.Builder.ComposerBot.Json
+{
+    public class SkillConversationIdFactory : SkillConversationIdFactoryBase
+    {
+        private readonly ConcurrentDictionary<string, string> _conversationRefs = new ConcurrentDictionary<string, string>();
```
Comment (gabog) [inline BotProject/Templates/CSharp/SkillConversationIdFactory.cs]: This is in memory, you probably want to use ConversationState or some other persistent state

**Comment by luhan2017** [comment]
We had a sync up with Gabo, thanks @gabog's help on reviewing this PR.
Here are some notes:
1.  confirmed my three questions above, MicrosoftAppId and MicrosoftPassword is required for both parent and child bot, because of the authentication.  We requested to remove the dependency on them on local scenarios, Gabo will keep this in mind.

2. We've got some comments from Gabo, Qi is goding to fix those.

3. This implementation is based on the first version of skill implementation and the bot to bot sample, it will keep updated. We have agreed that we implement this version in the current phase, and can update if there are further changes.

4. We need to think about the dialog to dialog mode, and skill manifest to handle more sotisphicated scenarios.

**Comment by luhan2017** [comment]
"SkillHostEndpoint": "http://localhost:3979/", // the endpoint where main bot hosted on

Let's remove the SkillHostEndpoint in the beginSkill action, it is a bot level setting, should set it in appsettings.json

**Comment by zidaneymar** [comment]
> "SkillHostEndpoint": "http://localhost:3979/", // the endpoint where main bot hosted on
> 
> Let's remove the SkillHostEndpoint in the beginSkill action, it is a bot level setting, should set it in appsettings.json

done

**Comment by luhan2017** [comment]
We need to verify the OAuth scenario.

**Comment by luhan2017** [comment]
Comments:
For this PR
1. Do we need "targetSkill" in schema now? can we make it just Id, appId and skillEndpoint? 
2. Can we make this template into another folder? Then we will have normal template and skill template, skill template is a specific template for custom schema ans custom action. 

New feature requirements for Composer after this PR is merged
1. SkillHostEndPoint which is set in the caller's appsetting.json should be auto-generated.
2. Composer need to support start multiple runtime instances so that we can test the skill feature in UI. Currently we can only test it in code level.
3. Similar as 2, Composer need to support create a bot from which template, (normal template or a custom template)
4. In this PR, we put the new schema in template/schema folder, Composer should support loading the schema then the visual and form part will also work. (This is the feature for custome schema and custom action.) Potentially, User need to configure some UI schema besides the runtime schema).

Next step for Skill 2.1
We are currently investigating to enable skill manifest in Composer, will update after holiday.

**Comment by zidaneymar** [comment]
> Comments:
> For this PR
> 
> 1. Do we need "targetSkill" in schema now? can we make it just Id, appId and skillEndpoint?
> 2. Can we make this template into another folder? Then we will have normal template and skill template, skill template is a specific template for custom schema ans custom action.
> 
> New feature requirements for Composer after this PR is merged
> 
> 1. SkillHostEndPoint which is set in the caller's appsetting.json should be auto-generated.
> 2. Composer need to support start multiple runtime instances so that we can test the skill feature in UI. Currently we can only test it in code level.
> 3. Similar as 2, Composer need to support create a bot from which template, (normal template or a custom template)
> 4. In this PR, we put the new schema in template/schema folder, Composer should support loading the schema then the visual and form part will also work. (This is the feature for custome schema and custom action.) Potentially, User need to configure some UI schema besides the runtime schema).
> 
> Next step for Skill 2.1
> We are currently investigating to enable skill manifest in Composer, will update after holiday.

1. For targetSkill, it's a temporary placeholder for skill manifest, currently we use id, appid and endpoint to represent manifest.
2. Actually the template doesn't influence the current runtime, we only add a customized step and skill controller inside the template.

**Comment by yeze322** [comment]
@zidaneymar Do you know what's the next step of this PR?

**Pull request closed without merging**